cpp/support/thread_safe_cache.h

```diff
@@ -179,6 +188,228 @@ class ThreadSafeCache<Key, Value> {
+    return entry.value;
+  }
+
+  template <typename Key>
+  auto lru_init(std::pair<const Key, Entry>& pair, const Value& init) -> void {
+    static_assert((std::is_same_v<Key, Keys> || ...), "Key type not found in the cache");
+    auto& [key, entry] = pair;
+    entry.value = init;
+    entry.iterator = lru_list_.emplace(lru_list_.end(), KeySet{&key}, entry);
+  }
+
+  template <typename Predicate, typename Evict>
+  auto lru_evict(const Predicate& predicate, const Evict& evict) -> void {
+    if (!predicate())  // short circuit if the predicate is false
+      return;
+
+    auto iter = lru_list_.begin();
+    if (iter == lru_list_.end()) return;
+
+    auto waiting = LRUlist{};
+
+    do {
+      auto& [keyset, entry] = *iter;
+      if (evict(entry.value)) {
+        std::visit(
+            [&](const auto* key) {
+              using Key_t = std::decay_t<decltype(*key)>;
+              get_map<Key_t>().erase(*key);
+            },
+            keyset
+        );
+        iter = lru_list_.erase(iter);
+      } else {
+        waiting.splice(waiting.end(), lru_list_, iter++);
+      }
+    } while (predicate() && iter != lru_list_.end());
+
+    // move the waiting list to the end of the lru list
+    lru_list_.splice(lru_list_.end(), waiting);
+  }
+
+ private:
+  std::tuple<std::unordered_map<Keys, Entry>...> storage_;
+  LRUlist lru_list_;
+};
+
+template <typename Value>
+struct SizedValue {
+  Value value;
+  std::size_t size;
+  template <typename Fn>
+  SizedValue(Value value, const Fn& size_fn) : value{value}, size{size_fn(value)} {}
+};
+
+template <typename Value, typename... Keys>
+using LRUCacheSizedImpl = LRUCacheImpl<std::shared_future<SizedValue<Value>>, Keys...>;
+
+template <typename Value>
+struct DemoPolicy {
+  // The interface of the policy
+  template <typename KeyType>
+  auto compute(const KeyType&) -> Value;
+};
+
+}  // namespace details
+
+/**
+ * \brief A thread-safe key-value cache with on-demand computation and LRU eviction
+ * \tparam Policy The policy class that provides the compute method
+ * \tparam Value The type of values stored in the cache
+ * \tparam Keys The types of keys used to lookup values. Should be hashable.
+ * \details This cache provides thread-safe access to computed values with the following features:
+ * - Lazy computation: Values are only computed when first requested
+ * - LRU eviction: When the cache is full, the least recently used value is evicted
+ * - Thread safety: Uses reader-writer locks for concurrent reads
+ * \attention User should guarantee the following:
+ * 1. The policy class should provide a compute method that takes a key and returns a value.
+ * 2. The value type should have a MemorySize method that returns the size of the value in bytes.
+ * 3. All the key types should be hashable by std::hash.
+ */
+template <typename Policy, typename Value, typename... Keys>
+class ThreadSafeLRUCache : private Policy, details::LRUCacheSizedImpl<Value, Keys...> {
```

Comment:
```suggestion
class ThreadSafeLRUCache : details::LRUCacheSizedImpl<Value, Keys...> {
```

always avoid Multiple inheritance